singer/src/main/java/com/pinterest/singer/writer/headersproviders/LoggingAuditHeadersProvider.java

```diff
@@ -0,0 +1,36 @@
+package com.pinterest.singer.writer.headersproviders;
+
+import com.pinterest.singer.common.SingerMetrics;
+import com.pinterest.singer.thrift.LogMessage;
+import com.pinterest.singer.writer.HeadersProvider;
+
+import com.twitter.ostrich.stats.Stats;
+import org.apache.kafka.common.header.Headers;
+import org.apache.thrift.TException;
+import org.apache.thrift.TSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoggingAuditHeadersProvider implements HeadersProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LoggingAuditHeadersProvider.class);
+
+  private static final String HEADER_KEY = "loggingAuditHeaders";
+  private static final TSerializer SER = new TSerializer();
```

Comment:
@zzhhhzz unfortunately TSerializer is not thread safe (ref: https://issues.apache.org/jira/browse/THRIFT-2218) therefore this design will likely need to be changed. Since you are creating a singleton the serialize method will likely lead to data corruption.

Reply:
@ambud  thanks for pointing out this issue early.  Glad to know there is a patch in the jira ticket your mentioned,  but unfortunately it never get any updated.   I will update the PR to make  each Writer has one LoggingAuditHeaderProvider instance to prevent possible thread unsafe issues.